### ISSUE TITLE
Failed to sign apk when exporting from Godot.app on MacOS

### ISSUE THREAD

**Issue by pgdanes** [godotengine/godot#81240]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Godot v4.1.1.stable - macOS 13.5.0 - Vulkan (Compatibility) - Apple M2 - Apple M2 (8 Threads)

### Issue description

Exporting android project fails to sign using apksigner when running godot from applications folder via Godot.app

 `All located 'apksigner' tools in /Users/pdanes/Library/Android/sdk/build-tools failed to execute`

However, when running godot executable from `/Applications/Godot.app/Contents/MacOS/Godot` it succeeds

### Steps to reproduce

**Using Application directory Godot.app**
1. Run Godot.app via Applications directory
2. Export android project via Project > Export > (select android export settings) > Export APK

Output:  
```
All located 'apksigner' tools in /Users/pdanes/Library/Android/sdk/build-tools failed to execute
```

Above also happens when using One Click Deploy

**Using the terminal**
1. Run godot from terminal by opening /Applications/Godot.app/Contents/MacOS/Godot
2. Select same project
3. Export android project via Project > Export > (select android export settings) > Export APK

Output:
```
Signing binary using: 
/Users/pdanes/Library/Android/sdk/build-tools/33.0.2/apksigner sign --verbose --ks <REDACTED> --ks-pass pass:<REDACTED> --ks-key-alias <REDACTED> /path/to/apk/export/game.apk
Signed
```

### Minimal reproduction project

N/A

I tested with a blank project setup to the bare minimum of requirements to export to android and have the same issue.

**Comment by pgdanes** [comment]
This seems to be due to path variable that Godot opens with by default

Running Godot.app through Finder/Spotlight and then executing
```GDScript
print_debug(OS.get_environment("path))
```
outputs:
```bash
/usr/bin:/bin:/usr/sbin:/sbin
```

Whereas, running through terminal would pick up my entire path variable. Most importantly `/opt/homebrew/opt/openjdk/bin`

**Comment by akien-mga** [comment]
That's by design when running from Finder, macOS doesn't give access to shell environment stuff.

I guess there might be a way to register this in the global PATH for all applications?

But AFAIK it's a macOS intentional limitation and not a Godot bug. CC @bruvzg

**Comment by bruvzg** [comment]
> But AFAIK it's a macOS intentional limitation and not a Godot bug.

It is a macOS limitation, apps ignore all shell environment variables, there's no way to change it globally. But an app can change its environment itself. So I guess we can add common paths like hombrew and macports to the PATH or add an editor setting to do it manually (trying to parse shell configs is probably gonna be overkill).

**Comment by bruvzg** [comment]
> Whereas, running through terminal would pick up my entire path variable, most importantly: /opt/homebrew/opt/openjdk/bin

Usually, JDK on macOS should be installed in `/Library/Java/JavaVirtualMachines` and operate via `/usr/bin/java` wrapper, so it's not dependent on PATH. I guess in this particular case, it's probably an issue with incomplete JDK install.

**Comment by pgdanes** [comment]
> > Whereas, running through terminal would pick up my entire path variable, most importantly: /opt/homebrew/opt/openjdk/bin
> 
> Usually, JDK on macOS should be installed in `/Library/Java/JavaVirtualMachines` and operate via `/usr/bin/java` wrapper, so it's not dependent on PATH. I guess in this particular case, it's probably an issue with incomplete JDK install.

Yeah this seems the case, reinstalling through brew outputs this line as help afterwards: 

```
For the system Java wrappers to find this JDK, symlink it with
  sudo ln -sfn /opt/homebrew/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk
 ```

Ran the command provided by homebrew and retested Godot.app opened from Finder and it exports fine.

It might still be worth adding homebrew/other package manager paths in the editors environment? Or expanding on the error from apksigner (if it even reports any more info)?

**Comment by manake** [comment]
I have this on Windows 11, so it's not just on macOS. The APK builds successfully but the error is present.

<img width="1276" alt="tt" src="https://github.com/godotengine/godot/assets/6874031/fbe10ae6-0e96-488b-8185-13a6a530bcd5">

**Comment by Alex2782** [comment]
@manake did you install the JDK correctly via .msi file? I might try it on Windows 10 later for comparison.

https://docs.godotengine.org/en/latest/tutorials/export/exporting_for_android.html
https://adoptium.net/de/temurin/releases/?variant=openjdk11&os=windows


for macOS there is a solution if you do not install it as described in the doc. This is badly comparable with Windows.
```
For the system Java wrappers to find this JDK, symlink it with
  sudo ln -sfn /opt/homebrew/opt/openjdk/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk.jdk
```

**Comment by manake** [comment]
@Alex2782
1. But JDK has no purpose other than generating `debug.keystore` via `keytool.exe`?
    - It's the SDK from Android Studio that does 99% of the job and I have all the Android Studio modules installed exactly as listed in the documentation, each with the correct version.
    - The error is in the SDK: `All located 'apksigner' tools in C:/Users/x/AppData/Local/Android/Sdk/build-tools failed to execute`
2. I set `JAVA_HOME` to OpenJDK 11.0.14.1 from `D:\Unity\2023.1.12f1\Editor\Data\PlaybackEngines\AndroidPlayer\OpenJDK` before but I installed the one from your link now and updated `JAVA_HOME` and it didn't change anything, the error remains the same.

BTW:
1. Android Studio Giraffe (2022.3.1) comes with JBR-17.0.6 currently, and currently we are in "upgrade to v17 now" phase: https://github.com/whichjdk/whichjdk.com
2. Secondly, IMO there is no point in even having "build APK" from within GoDot:
    - Unity has "Unity as a Library" (https://unity.com/features/unity-as-a-library), which exports a Gradle project that you can extend and compile in Android Studio.
    - Unity also added `GameActivity` instead of `Activity` just a few weeks ago. Do you have this? If not, then I would focus on this and not on fixing the issue from this thread.
    - For iOS, Unity only has "export Xcode project", and everyone is fine with that.
    - GoDot/Unity/Unreal always include outdated Android Studio libraries (by 6-12 months), which is a problem. For example, now we should be at API v34 but Unity and GoDot are at v32 and v33. Most libraries that these engines include are always outdated, too. (And you can't actually include any AARs because they will conflict with the Android Studio ones, so any "Android Plugins" for GoDot are not going to happen anyway because the only right place to load them is in Android Studio.)
    - By removing building APK functionality from GoDot you would remove much unnecessary code to maintain, you would get rid of many bugs, and no functionality will be lost for users because they can build it in Android Studio.
    - **I'm coming from Android Studio environment and I don't even understand why would GoDot/Unity/Unreal ever create "build APK" functionality in the first place. It just doesn't make any sense to reinvent the wheel and duplicate the functionality that Android Studio has and does better. So, all these "specify app icon" and "select permissions" UI fields in GoDot are completely unnecessary IMO. You can just remove them and that will be only benefit. IMO, you should just focus on generating a gradle project with `GoDotGameActivity.java` file with C++ `canvas` inside and provide `GoDot <-> Android Studio` communication bridge and just remove "build APK". Your "build APK" doesn't even allow integrating AdMob or Google Play Services and many other Android libraries (unless by including AAR into GoDot via https://github.com/Poing-Studios/godot-admob-android which is not how you normally would do it, this is a workaround really that will soon have outdated AAR), so a Gradle project is necessary for most mobile developers anyway.**

I could go on but the point is:
- There are only benefits and no drawbacks of this decision.
- GoDot/Unity/Unreal users are not stupid and we can build gradle project in Android Studio, edit manifest.xml etc.
- Please think about this, it looks like 5-10 major benefits and 0 drawbacks if GoDot would remove the many non-essential features and instead limit the scope to only delivering core "gradle project export" feature.
- No need to reinvent the wheel and have "build APK" functionality in GoDot that is useless for Android Studio users who only want gradle project and they don't trust/want APKs built by other tools like GoDot/Unity/Unreal.

**Comment by Calinou** [comment]
Please don't derail the discussion on whether this feature is needed or not. We need to fix the issue regardless, or improve the documentation.

**Comment by manake** [comment]
> Please don't derail the discussion on whether this feature is needed or not.

I agree. Let's not derail the discussion on whether this feature is needed or not. 👌

**Comment by phaitonican** [comment]
this is also on linux btw

**Comment by phaitonican** [comment]
same error on my nixos either maybe i do something wrong idk
